pkg/controller/expcertificates/issuing/issuing_controller.go

```diff
@@ -0,0 +1,396 @@
+type secretData struct {
+	sk, cert, ca []byte
+}
+
+func NewController(
+	log logr.Logger,
+	client cmclient.Interface,
+	factory informers.SharedInformerFactory,
+	cmFactory cminformers.SharedInformerFactory,
+	recorder record.EventRecorder,
+	clock clock.Clock,
+) (*controller, workqueue.RateLimitingInterface, []cache.InformerSynced) {
+	// create a queue used to queue up items to be processed
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(time.Second*1, time.Second*30), ControllerName)
+
+	// obtain references to all the informers used by this controller
+	certificateInformer := cmFactory.Certmanager().V1alpha2().Certificates()
+	certificateRequestInformer := cmFactory.Certmanager().V1alpha2().CertificateRequests()
+	secretsInformer := factory.Core().V1().Secrets()
+
+	certificateInformer.Informer().AddEventHandler(&controllerpkg.QueuingEventHandler{Queue: queue})
+	certificateRequestInformer.Informer().AddEventHandler(&controllerpkg.BlockingEventHandler{
+		WorkFunc: controllerpkg.HandleOwnedResourceNamespacedFunc(log, queue, certificateGvk, certificates.CertificateGetFunc(certificateInformer.Lister())),
+	})
+
+	// build a list of InformerSynced functions that will be returned by the Register method.
+	// the controller will only begin processing items once all of these informers have synced.
+	mustSync := []cache.InformerSynced{
+		certificateRequestInformer.Informer().HasSynced,
+		secretsInformer.Informer().HasSynced,
+		certificateInformer.Informer().HasSynced,
+	}
+
+	return &controller{
+		certificateLister:        certificateInformer.Lister(),
+		certificateRequestLister: certificateRequestInformer.Lister(),
+		secretLister:             secretsInformer.Lister(),
+		client:                   client,
+		recorder:                 recorder,
+		clock:                    clock,
+
+		//enableSecretOwnerReferences: ctx.CertificateOptions.EnableOwnerRef
+	}, queue, mustSync
+}
+
+func (c *controller) ProcessItem(ctx context.Context, key string) error {
+	log := logf.FromContext(ctx).WithValues("key", key)
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil
+	}
+
+	crt, err := c.certificateLister.Certificates(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		log.Error(err, "certificate not found for key")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if !apiutil.CertificateHasCondition(crt, cmapi.CertificateCondition{
+		Type:   cmapi.CertificateConditionIssuing,
+		Status: cmmeta.ConditionTrue,
+	}) {
+		// Do nothing if an issuance is not in progress.
+		return nil
+	}
+
+	if crt.Status.NextPrivateKeySecretName == nil ||
+		len(*crt.Status.NextPrivateKeySecretName) == 0 {
+		// Do nothing if the next private key secret name is not set
+		return nil
+	}
+
+	nextPrivateKeySecretName := *crt.Status.NextPrivateKeySecretName
+
+	reqs, nextRevision, err := findNextRevisionCertificateRequestsForCertificate(log, crt, c.certificateRequestLister)
```

Comment:
Similar to the `trigger` controller, you should be able to remove this function altogether and instead use:

```
		reqs, err := certificates.ListCertificateRequestsMatchingPredicate(c.certificateRequestLister.CertificateRequests(crt.Namespace),
			labels.Everything(),
			certificates.WithCertificateRevisionPredicateFunc(*crt.Status.Revision),
		)
```

(replacing `*crt.Status.Revision` with an appropriate integer revision)

Reply:
You may also want to add a new 'predicate function' that ensures the owner of the certificaterequest is the certificate currently being processed too. `ListCertificateRequestsMatchingPredicate` can then be modified to accept `predicate` as a variadic (i.e. `predicates ...CertificateRequestPredicateFunc`)